### PR TITLE
docs(ref): Provide guidance on version requirements

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

I've been dealing with these situations as either the package author, depending on such a package, or supporting users who run into problems that I figure documenting this guidance in a central place means I won't have to repeat myself as often and have to re-find all of the relevant links each time.

### How should we test and review this PR?

Alternatives to how this was documented
- Use a regular header.  All of sections in this document are flat and its hard to see association between them.  This also feels like its more on the level of the "note"s.
- Put this in a central Recommendations page.  I think we should do something more for these when we have more (nothing else in my quick scan stood out as "recommendations" like this).  At that point we can have a better idea of how it would work (much like the rule of 3 for generalizing code).  I also suspect a "Recommendations" index might be better.
- Put this in the FAQ.  This can easily be framed as a question and we put the `Cargo.lock` policy in there.

I left out talking about alternative proc-macro designs as I feel like treading new ground in community practices is out of the scope of this.

### Additional information

See also https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Version.20Requirements.20documentation

### CODE REVIEW

File: src/doc/src/reference/specifying-dependencies.md [hunk 1]
```diff
@@ -107,6 +107,30 @@ Here are some examples of comparison requirements:
...
+>
+> Avoid constraining the upper bound of a version to be anything less than the next semver incompatible version
+> (e.g. avoid `">=2.0, <2.4"`) as other packages in the dependency tree may require a newer version, leading to an unresolvable error (see #6584).
+> Consider whether controlling the version in your [`Cargo.lock`] would be more appropriate.
+>
+> In some instances this won't matter or the benefits might outweigh the cost, including:
+> - When no one else depends on your package e.g. it only has a `[[bin]]`
+> - When depending on a pre-release package and wishing to avoid breaking changes then a fully specified `"=1.2.3-alpha.3"` might be warranted (see [#2222])
+> - When a crate re-exports a proc-macro but the proc-macro generates code that calls into the re-exporting crate then a fully specified `=1.2.3` might be warranted to ensure the proc-macro isn't newer than the re-exporting crate and generating code that uses parts of the API that don't exist within the current version
```
Comment (weihanglo) [inline src/doc/src/reference/specifying-dependencies.md:127]: Would it be better if we use `library` here?

```suggestion
> - When a library re-exports a proc-macro but the proc-macro generates code that calls into the re-exporting library then a fully specified `=1.2.3` might be warranted to ensure the proc-macro isn't newer than the re-exporting library and generating code that uses parts of the API that don't exist within the current version
```

**Review by weihanglo — COMMENTED** [review]
Read through the Zulip thread, I think people generally agree on these recommendations. I'll merge it without FCP once the minor suggestions are resolved.

File: src/doc/src/reference/specifying-dependencies.md [hunk 2]
```diff
@@ -107,6 +107,30 @@ Here are some examples of comparison requirements:
...
 
+> **Recommendation:** When in doubt, use the default version requirement operator.
+>
+> In rare circumstances, a package with a "public dependency"
+> (re-exports the dependency or interoperates with it in its public API)
+> that is compatible with multiple semver-incompatible versions
+> (e.g. only uses a simple type that hasn't changed between releases, like an `Id`)
+> may support users choosing which version of the "public dependency" to use.
+> In this case, a version requirement like `">=0.4, <2"` may be of interest.
+> *However* users of the package will likely run into errors and need to to manually select a version of the "public dependency" via `cargo update` if they also depend on it as Cargo might pick different versions of the "public dependency" when [resolving dependency versions][resolver.md]  (see [#10599]).
```
[the remark this answers is not in the document]
Comment (epage) [reply src/doc/src/reference/specifying-dependencies.md:118]: My wrapping was actually on sentence (or parenthetical) breaks but I can do some more wrapping.

I've been finding its nicer to at least wrap on end-of-sentences as it makes it easier to see the diff as the whole thing doesn't get re-wrapped on changes that makes it impossible to see what changed.

[the remark this answers is not in the document]
Comment (weihanglo) [reply src/doc/src/reference/specifying-dependencies.md]: Understand! I am not sure which way to go. There are at least three options:

* Hard warp on 80 or 120 columns.
* [Semantic linefeeds](https://rhodesmill.org/brandon/2012/one-sentence-per-line/). See <https://github.com/rust-lang/cargo/pull/10916#discussion_r934132791>
* Wrap only at the end of sentences.

**Review by weihanglo — APPROVED** [review]

File: src/doc/src/reference/specifying-dependencies.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (epage) [reply src/doc/src/reference/specifying-dependencies.md]: Oh hey, I'm not the only one who has been moving in that direction

**Pull request merged**